drivers/mock/driver.go

```diff
@@ -0,0 +1,441 @@
+	}
+
+	if driverConfig.StartErr != "" {
+		return nil, nil, structs.NewRecoverableError(errors.New(driverConfig.StartErr), driverConfig.StartErrRecoverable)
+	}
+
+	// Create the driver network
+	net := &cstructs.DriverNetwork{
+		IP:            driverConfig.DriverIP,
+		AutoAdvertise: driverConfig.DriverAdvertise,
+	}
+	if raw := driverConfig.DriverPortMap; len(raw) > 0 {
+		parts := strings.Split(raw, ":")
+		if len(parts) != 2 {
+			return nil, nil, fmt.Errorf("malformed port map: %q", raw)
+		}
+		port, err := strconv.Atoi(parts[1])
+		if err != nil {
+			return nil, nil, fmt.Errorf("malformed port map: %q -- error: %v", raw, err)
+		}
+		net.PortMap = map[string]int{parts[0]: port}
+	}
+
+	killCtx, killCancel := context.WithCancel(context.Background())
+
+	h := &mockTaskHandle{
+		task:            cfg,
+		runFor:          driverConfig.RunFor,
+		killAfter:       driverConfig.KillAfter,
+		exitCode:        driverConfig.ExitCode,
+		exitSignal:      driverConfig.ExitSignal,
+		stdoutString:    driverConfig.StdoutString,
+		stdoutRepeat:    driverConfig.StdoutRepeat,
+		stdoutRepeatDur: driverConfig.StdoutRepeatDur,
+		logger:          d.logger.With("task_name", cfg.Name),
+		waitCh:          make(chan struct{}),
+		killCh:          killCtx.Done(),
+		kill:            killCancel,
+	}
+	if driverConfig.ExitErrMsg != "" {
+		h.exitErr = errors.New(driverConfig.ExitErrMsg)
+	}
+	if driverConfig.SignalErr != "" {
+		h.signalErr = fmt.Errorf(driverConfig.SignalErr)
+	}
+
+	driverState := MockTaskState{
+		TaskConfig: cfg,
+		StartedAt:  h.startedAt,
+	}
+	handle := drivers.NewTaskHandle(pluginName)
+	handle.Config = cfg
+	if err := handle.SetDriverState(&driverState); err != nil {
+		d.logger.Error("failed to start task, error setting driver state", "error", err)
+		return nil, nil, fmt.Errorf("failed to set driver state: %v", err)
+	}
+
+	d.tasks.Set(cfg.ID, h)
+
+	d.logger.Debug("starting task", "name", cfg.Name)
+	go h.run()
+	return handle, net, nil
+
+}
+
+func (d *Driver) WaitTask(ctx netctx.Context, taskID string) (<-chan *drivers.ExitResult, error) {
+	handle, ok := d.tasks.Get(taskID)
+	if !ok {
+		return nil, drivers.ErrTaskNotFound
+	}
+
+	ch := make(chan *drivers.ExitResult)
+	go d.handleWait(ctx, handle, ch)
+
+	return ch, nil
+
+}
+func (d *Driver) handleWait(ctx context.Context, handle *mockTaskHandle, ch chan *drivers.ExitResult) {
+	defer close(ch)
+
+	select {
+	case <-ctx.Done():
+		return
+	case <-d.ctx.Done():
+		return
+	case <-handle.waitCh:
+		ch <- handle.exitResult
+	}
+}
+func (d *Driver) StopTask(taskID string, timeout time.Duration, signal string) error {
+	h, ok := d.tasks.Get(taskID)
+	if !ok {
+		return drivers.ErrTaskNotFound
+	}
+
+	d.logger.Debug("killing task",
+		"task_name", h.task.Name,
+		"kill_after", h.killAfter,
+		"kill_timeout", h.killTimeout,
+	)
+
+	select {
+	case <-h.waitCh:
+		d.logger.Debug("not killing task: already exited", "task_name", h.task.Name)
+	case <-time.After(h.killAfter):
```

Comment:
This doesn't make sense. It is waiting on unrelated timeouts.

Reply:
Right? I just ported the existing code and it confused me too: https://github.com/hashicorp/nomad/blob/master/client/driver/mock_driver.go#L380-L402

Should I do a grep and switch everything to `kill_timeout` and remove `kill_after`?

Reply:
@nickethier pointed out killTimeout isn't used. Removing that one.

Reply:
Ah I think I see. If not set the kill will happen immediately. So it is just a way to introduce a kill delay.